core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,169 @@
+public class PushMergeWriterUpdateIntoConnector
+        implements Rule<TableFinishNode>
+{
+    private static final Capture<MergeWriterNode> MERGE_WRITER_NODE_CAPTURE = newCapture();
+    private static final Capture<MergeProcessorNode> MERGE_PROCESSOR_NODE_CAPTURE = newCapture();
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+    private static final Capture<ProjectNode> PROJECT_NODE_CAPTURE = newCapture();
+
+    private static final Pattern<TableFinishNode> PATTERN =
+            tableFinish().with(source().matching(
+                    mergeWriter().capturedAs(MERGE_WRITER_NODE_CAPTURE).with(source().matching(
+                            mergeProcessor().capturedAs(MERGE_PROCESSOR_NODE_CAPTURE).with(source().matching(
+                                    project().capturedAs(PROJECT_NODE_CAPTURE).with(source().matching(
+                                            tableScan().capturedAs(TABLE_SCAN)))))))));
+
+    public PushMergeWriterUpdateIntoConnector(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer, Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    private final Metadata metadata;
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<TableFinishNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TableFinishNode node, Captures captures, Context context)
+    {
+        MergeWriterNode mergeWriter = captures.get(MERGE_WRITER_NODE_CAPTURE);
+        MergeProcessorNode mergeProcessor = captures.get(MERGE_PROCESSOR_NODE_CAPTURE);
+        ProjectNode project = captures.get(PROJECT_NODE_CAPTURE);
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(context.getSession(), mergeWriter.getTarget().getHandle());
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(mergeProcessor.getMergeRowSymbol()).getChildren();
+
+        Map<ColumnHandle, Constant> assignments = buildAssignments(orderedColumnNames, mergeAssignments, columnHandles, context);
+        if (assignments.isEmpty()) {
+            return Result.empty();
+        }
+
+        // Query Planner builds the merge rowblock, containing:
+        // All data columns in table order
+        // plus 3 additional fields: field present flag, update operation number, merge case number
+        // TODO move this check before buildAssignments as soon support for update all will be added,
+        //  in case of update all - row encoded as base64, which filtered in buildAssignments
```

Comment:
While it is true that it's encoded using base64, it's irrelevant for this code. What matters is that it's represented as a function call. Remove this comment.